modules/videoio/include/opencv2/videoio/utils.private.hpp

```diff
@@ -10,6 +10,83 @@
 
 namespace cv {
 CV_EXPORTS std::string icvExtractPattern(const std::string& filename, unsigned *offset);
+
+class IReadStream
+{
+public:
+    virtual ~IReadStream() {};
+    virtual long long read(char* buffer, long long size) = 0;
+    virtual long long seek(long long offset, int way) = 0;
+    virtual IReadStream* clone() = 0;
+};
+
+class StreambufReadStream : public IReadStream
+{
+public:
+    StreambufReadStream(std::streambuf& _stream) : stream(_stream) {}
+
+    virtual ~StreambufReadStream() {}
+
+    long long read(char* buffer, long long size) override
+    {
+        return stream.sgetn(buffer, size);
+    }
+
+    long long seek(long long offset, int way) override
+    {
+        return stream.pubseekoff(offset, way == SEEK_SET ? std::ios_base::beg : (way == SEEK_END ? std::ios_base::end : std::ios_base::cur));
+    }
+
+    IReadStream* clone() override
+    {
+        return new StreambufReadStream(stream);
+    }
+
+    static Ptr<IReadStream> create(std::streambuf& stream)
+    {
+        return Ptr<IReadStream>(new StreambufReadStream(stream));
+    }
+
+private:
+    std::streambuf& stream;
+};
+
+
+class ReadStreamCallback : public IReadStream
```

Comment:
`ReadStreamPluginProvider`

Usually `Callback` means abstract interface description. We already have the `IReadStream` for that.